### ISSUE TITLE
substitute_project_variables returns incorrect results with overrrides

### ISSUE THREAD

**Issue by kinghuang** [explosion/spaCy#7736]
I'm trying to pass a version value like `0.0.0` as a command override to `spacy project run …`. Values like `0` and `0.0` work, but `0.0.0` produces `TypeError: string indices must be integers`.

```
root@738ace767785:/project/lib# spacy project run package-dist . --vars.version=0.1.1d
Traceback (most recent call last):
  File "/usr/local/bin/spacy", line 8, in <module>
    sys.exit(app())
  File "/usr/local/lib/python3.8/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/usr/local/lib/python3.8/site-packages/spacy/cli/project/run.py", line 41, in project_run_cli
    project_run(project_dir, subcommand, overrides=overrides, force=force, dry=dry)
  File "/usr/local/lib/python3.8/site-packages/spacy/cli/project/run.py", line 69, in project_run
    commands = {cmd["name"]: cmd for cmd in config.get("commands", [])}
  File "/usr/local/lib/python3.8/site-packages/spacy/cli/project/run.py", line 69, in <dictcomp>
    commands = {cmd["name"]: cmd for cmd in config.get("commands", [])}
TypeError: string indices must be integers
```

## How to reproduce the behaviour

It seems that under some conditions, [spacy.cli._util.substitute_project_variables()](https://github.com/explosion/spaCy/blob/673e2bc4c0ca46f1c026e0823d32f35c52d2f38e/spacy/cli/_util.py#L171) returns incorrectly interpolated configs when overrides are used. Here's the code from that function, modified to demonstrate the problem. `greeting` is a variable that will be overridden.

```python
from thinc.api import Config

config = {
  "vars": {"greeting": "hello"},
  "commands": [{"name": "greet", "script": ["echo ${vars.greeting}"]}],
  "env": {}
}
overrides = {"vars.greeting": "bonjour"}

cfg = Config({"project": config, "vars": config["vars"], "env": config["env"]})

cfg1 = Config().from_str(cfg.to_str())
cfg2 = Config().from_str(cfg.to_str(), overrides=overrides)
```

**cfg1.to_str():**

`cfg1` doesn't have overrides, and its result looks normal.

```toml
[env]

[project]
commands = [{"name":"greet","script":["echo hello"]}]

[project.env]

[project.vars]
greeting = "hello"

[vars]
greeting = "hello"
```

**cfg2.to_str():**

`cfg2` has overrides. Notice that `commands` is now a string form of the command list instead of an actual list. Extra quotes also appeared for `bonjour`.

```toml
[env]

[project]
commands = "[{\"name\":\"greet\",\"script\":[\"echo \"bonjour\"\"]}]"

[project.env]

[project.vars]
greeting = "hello"

[vars]
greeting = "bonjour"
```

This originated from a question I posted: https://github.com/explosion/spaCy/discussions/7734

## Your Environment

* Operating System: Debian GNU/Linux 10 (buster)
* Python Version Used: 3.8.9
* spaCy Version Used: 3.0.5
* Environment Information: Docker python container

**Comment by svlandeg** [comment]
Hi King,

You're right that it seems like there's an issue here with non-numeric values, like `"bonjour"` or `1.2.3`. We'll look into it!

**Comment by kinghuang** [comment]
That's a fair point about it being undocumented functionality! I'm using `env` (also undocumented) as a workaround right now.

Both are very useful features for my workflow. 😃

**Comment by svlandeg** [comment]
This should be fixed in the next release, probably 3.0.6. Thanks again for the report!
